megatron/data/ul2_dataset.py

```diff
@@ -0,0 +1,277 @@
+        target_seq_length: Desired sequence length.
+        max_seq_length: Maximum length of the sequence. All values are padded to
+            this length.
+        max_seq_length_dec: Maximum length of the decoder input sequence. All
+            values are padded to this length.
+        vocab_id_list: List of vocabulary ids. Used to pick a random id.
+        vocab_id_to_token_dict: A dictionary from vocab ids to text tokens.
+        cls_ids: Start of example ids.
+        sep_id: Separator id.
+        mask_id: Mask token id.
+        pad_id: Padding token id.
+        model_type: What type of model is used.
+        denoiser_ratios: Probability of each denoising objective to be selected.
+        denoisers: What type of UL2 denoising objective the other UL2
+              configurations refer to.
+        mean_span_lengths: Mean length for sampling span lengths. Numbers < 1
+              indicate a mean length of the sequence length times that number.
+        mask_ratios: Ratio of masked token in the full sequence.
+        np_rng: Random number genenrator. Note that this rng state should be
+              numpy and not python since python randint is inclusive for
+              the opper bound whereas the numpy one is exclusive.
+        bos_id: start of decoder example id
+        eos_id: end of generation id
+        sentinel_tokens: unique value to be substituted for every replaced span
+    """
+
+    # Denoiser selection
+    denoiser_index = np_rng.choice(np.arange(len(denoisers)), p=denoiser_ratios)
+    denoiser = denoisers[denoiser_index]
+    masked_lm_prob = mask_ratios[denoiser_index]
+
+    assert target_seq_length <= max_seq_length
+
+    # flatten sentences into one list
+    tokens = [token for sentence in sample for token in sentence]
+
+    max_num_tokens = target_seq_length
+    # if is_decoder_only(model_type):
+    #     # Keep space for repeated `extra_id` tokens; not the most data
+    #     # efficient since we calculate this based on the maximum number
+    #     # of possible `extra_id` tokens.
+    #     safe_max_seq_len = math.floor(max_num_tokens / (1 + masked_lm_prob))
+    #     truncated = len(tokens) > safe_max_seq_len
+    #     tokens = tokens[:safe_max_seq_len]
+    # else:
+    # Truncate to `target_sequence_length`.
+    truncated = len(tokens) > max_num_tokens
+    tokens = tokens[:max_num_tokens]
+
+    # Prepend objective token.
+    cls_id = cls_ids.get(denoiser)
+    if cls_id is None:
+        raise ValueError('unknown denoiser')
+    tokens = [cls_id] + tokens
+
+    # Masking.
+    # Ensure we always have at least one prediction.
+    max_predictions_per_seq = max(1.0, masked_lm_prob * len(tokens))
+    mean_ngrams = mean_span_lengths[denoiser_index]
+    if mean_ngrams < 1:
+        # Ensure we always obtain at least one `max_ngrams`.
+        mean_ngrams = max(1, round(len(tokens) * mean_ngrams))
+    max_ngrams = mean_ngrams * 2 - 1
```

Comment:
Assume we have 10 tokens in the sequence and `mean_ngrams` is 2, i.e. we would like to have a probability distribution centered on 2 with 1 as the minimum value. We calculate the distance from 1 to `mean_ngrams = 2` and add it to our center `mean_ngrams` to get the inclusive upper bound. The doubling is simply the addition, here. We could/should equivalently write:
```python
distribution_radius = mean_ngrams - 1
max_ngrams = mean_ngrams + distribution_radius
```

On the one hand, this handles normal distributions because we calculate the distribution's mean like `normal_mean = (max_ngrams + 1) / 2` (exactly `mean_ngrams` again) inside `create_masked_lm_predictions`.
However, on the other hand, now I also see that I should modify [the `ngrams` calculation inside `create_masked_lm_predictions`](https://github.com/NVIDIA/Megatron-LM/pull/268/files/324d70d2518fe183c27e36ed550a95955eaa077c#diff-e1d14be32f4489a01cb8d571804fbba003f7f90715ef3cb3a27d9099e0245d6fR260) to remove the upper bound on `max_ngrams` if the sampling distribution does not have an upper bound.